File: 2.0.2/effects/channel_shift.py
# effects/channel_shift.py
import numpy as np
from PIL import Image

def apply_channel_shift(image, params, selections=None):
    img_array = np.array(image)
    height, width, _ = img_array.shape
    
    shift_r = int(params['shift_r'])
    shift_g = int(params['shift_g'])
    shift_b = int(params['shift_b'])
    axis_r = params['axis_r']
    axis_g = params['axis_g']
    axis_b = params['axis_b']
    
    shifted_array = img_array.copy()
    
    if selections:
        for start, end, channel in selections:
            if channel == 0:  # Red channel
                shifted_array[:, start:end, 0] = apply_shift(img_array[:, start:end, 0], shift_r, axis_r)
            elif channel == 1:  # Green channel
                shifted_array[:, start:end, 1] = apply_shift(img_array[:, start:end, 1], shift_g, axis_g)
            elif channel == 2:  # Blue channel
                shifted_array[:, start:end, 2] = apply_shift(img_array[:, start:end, 2], shift_b, axis_b)
    else:
        # Apply to entire image if no selections
        shifted_array[:, :, 0] = apply_shift(img_array[:, :, 0], shift_r, axis_r)
        shifted_array[:, :, 1] = apply_shift(img_array[:, :, 1], shift_g, axis_g)
        shifted_array[:, :, 2] = apply_shift(img_array[:, :, 2], shift_b, axis_b)
    
    return Image.fromarray(shifted_array)

def apply_shift(channel, shift, axis):
    if axis == 'horizontal':
        return np.roll(channel, shift, axis=1)
    else:  # vertical
        return np.roll(channel, shift, axis=0)

File: webapp/static/js/effects/channel_shift.js
// webapp/static/js/effects/channel_shift.js
'use strict';

/**
 * Apply channel shift effect to image data.
 * Shifts RGB channels independently along horizontal or vertical axes.
 * 
 * @param {ImageData} imageData - The source image data
 * @param {Object} params - Effect parameters
 * @param {number} params.shift_r - Red channel shift percentage (-100 to 100)
 * @param {string} params.axis_r - Red shift axis ('horizontal' or 'vertical')
 * @param {number} params.shift_g - Green channel shift percentage (-100 to 100)
 * @param {string} params.axis_g - Green shift axis ('horizontal' or 'vertical')
 * @param {number} params.shift_b - Blue channel shift percentage (-100 to 100)
 * @param {string} params.axis_b - Blue shift axis ('horizontal' or 'vertical')
 * @returns {ImageData} The processed image data
 */
export function applyChannelShift(imageData, params) {
    const { width, height, data } = imageData;
    const result = new Uint8ClampedArray(data);
    
    // Parse and clamp parameters
    const shiftRPct = Math.max(-100, Math.min(100, params.shift_r || 0)) / 100;
    const shiftGPct = Math.max(-100, Math.min(100, params.shift_g || 0)) / 100;
    const shiftBPct = Math.max(-100, Math.min(100, params.shift_b || 0)) / 100;
    
    const axisR = params.axis_r === 'vertical' ? 'vertical' : 'horizontal';
    const axisG = params.axis_g === 'vertical' ? 'vertical' : 'horizontal';
    const axisB = params.axis_b === 'vertical' ? 'vertical' : 'horizontal';
    
    // Calculate pixel shifts
    const shiftR = Math.round(axisR === 'horizontal' ? width * shiftRPct : height * shiftRPct);
    const shiftG = Math.round(axisG === 'horizontal' ? width * shiftGPct : height * shiftGPct);
    const shiftB = Math.round(axisB === 'horizontal' ? width * shiftBPct : height * shiftBPct);
    
    // Apply shifts for each pixel
    for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
            const idx = (y * width + x) * 4;
            
            // Red channel
            const srcR = getShiftedPixelIndex(x, y, width, height, shiftR, axisR);
            result[idx] = data[srcR * 4];
            
            // Green channel
            const srcG = getShiftedPixelIndex(x, y, width, height, shiftG, axisG);
            result[idx + 1] = data[srcG * 4 + 1];
            
            // Blue channel
            const srcB = getShiftedPixelIndex(x, y, width, height, shiftB, axisB);
            result[idx + 2] = data[srcB * 4 + 2];
            
            // Alpha channel unchanged
            result[idx + 3] = data[idx + 3];
        }
    }
    
    return new ImageData(result, width, height);
}

/**
 * Get the source pixel index after applying shift with wrapping.
 */
function getShiftedPixelIndex(x, y, width, height, shift, axis) {
    let srcX = x;
    let srcY = y;
    
    if (axis === 'horizontal') {
        srcX = ((x - shift) % width + width) % width;
    } else {
        srcY = ((y - shift) % height + height) % height;
    }
    
    return srcY * width + srcX;
}
